# Module 4: Working with Text Data

Preprocessing our data is very important.

We may encounter data that is not clean. As we can see in DataFrame below:
- First name and last name are separated by a comma.
- Employee Annual Salary has a dollar ( $ ) sign, which means we cannot do mathematical operation on it.

In [1]:
import pandas as pd
df = pd.read_csv("data/chicago.csv")
df.head()

Name            Position Title        Department  \
0      AARON,  ELVIA J          WATER RATE TAKER       WATER MGMNT   
1    AARON,  JEFFERY M            POLICE OFFICER            POLICE   
2       AARON,  KARINA            POLICE OFFICER            POLICE   
3  AARON,  KIMBERLEI R  CHIEF CONTRACT EXPEDITER  GENERAL SERVICES   
4  ABAD JR,  VICENTE M         CIVIL ENGINEER IV       WATER MGMNT   

  Employee Annual Salary  
0              $90744.00  
1              $84450.00  
2              $84450.00  
3              $89880.00  
4             $106836.00

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32063 entries, 0 to 32062
Data columns (total 4 columns):
Name                      32062 non-null object
Position Title            32062 non-null object
Department                32062 non-null object
Employee Annual Salary    32062 non-null object
dtypes: object(4)
memory usage: 1002.0+ KB


In Department column, we have 35 categories. It is much better to change the column type to **category** as we have 32,063 rows of data. Rather than having 32,063 strings, we can classify them into 35 categories.

You can see that the memory usage decreases from 1002KB to 784.4KB. This may speed up our operation so much more.

In [3]:
df["Department"].nunique()

35

In [4]:
df["Department"] = df["Department"].astype("category")

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32063 entries, 0 to 32062
Data columns (total 4 columns):
Name                      32062 non-null object
Position Title            32062 non-null object
Department                32062 non-null category
Employee Annual Salary    32062 non-null object
dtypes: category(1), object(3)
memory usage: 784.4+ KB


### Remove null values

In [6]:
df.tail()

Name           Position Title Department  \
32058   ZYGOWICZ,  PETER J           POLICE OFFICER     POLICE   
32059    ZYMANTAS,  MARK E           POLICE OFFICER     POLICE   
32060  ZYRKOWSKI,  CARLO E           POLICE OFFICER     POLICE   
32061  ZYSKOWSKI,  DARIUSZ  CHIEF DATA BASE ANALYST       DoIT   
32062                  NaN                      NaN        NaN   

      Employee Annual Salary  
32058              $87384.00  
32059              $84450.00  
32060              $87384.00  
32061             $113664.00  
32062                    NaN

In [7]:
df.dropna(how="all", inplace=True)
df.tail()

Name                  Position Title        Department  \
32057  ZYGADLO,  MICHAEL J  FRM OF MACHINISTS - AUTOMOTIVE  GENERAL SERVICES   
32058   ZYGOWICZ,  PETER J                  POLICE OFFICER            POLICE   
32059    ZYMANTAS,  MARK E                  POLICE OFFICER            POLICE   
32060  ZYRKOWSKI,  CARLO E                  POLICE OFFICER            POLICE   
32061  ZYSKOWSKI,  DARIUSZ         CHIEF DATA BASE ANALYST              DoIT   

      Employee Annual Salary  
32057              $99528.00  
32058              $87384.00  
32059              $84450.00  
32060              $87384.00  
32061             $113664.00

### Python built-in function : lower(), upper(), title() , len()

- lower() : returns the string in lowercase. 
- upper() : returns a copy of the string in uppercase.
- title() : convert the first character in each word to uppercase and remaining characters to lowercase in string
- len() : calculate the length of string/ list/ etc. (In string, space also counts as a character)

In [8]:
"HELLO WORLD".lower()

'hello world'

In [9]:
"hello world".upper()

'HELLO WORLD'

In [10]:
"hello world".title()

'Hello World'

In [11]:
len("HELLO WORLD")

11

These functions allow us to change the format in Name column into something more readable.

In [12]:
df["Name"] = df["Name"].str.title()
df.head()

Name            Position Title        Department  \
0      Aaron,  Elvia J          WATER RATE TAKER       WATER MGMNT   
1    Aaron,  Jeffery M            POLICE OFFICER            POLICE   
2       Aaron,  Karina            POLICE OFFICER            POLICE   
3  Aaron,  Kimberlei R  CHIEF CONTRACT EXPEDITER  GENERAL SERVICES   
4  Abad Jr,  Vicente M         CIVIL ENGINEER IV       WATER MGMNT   

  Employee Annual Salary  
0              $90744.00  
1              $84450.00  
2              $84450.00  
3              $89880.00  
4             $106836.00

The same goes for Position Title column and Department column.

In [13]:
df["Position Title"] = df["Position Title"].str.title()
df.head()

Name            Position Title        Department  \
0      Aaron,  Elvia J          Water Rate Taker       WATER MGMNT   
1    Aaron,  Jeffery M            Police Officer            POLICE   
2       Aaron,  Karina            Police Officer            POLICE   
3  Aaron,  Kimberlei R  Chief Contract Expediter  GENERAL SERVICES   
4  Abad Jr,  Vicente M         Civil Engineer Iv       WATER MGMNT   

  Employee Annual Salary  
0              $90744.00  
1              $84450.00  
2              $84450.00  
3              $89880.00  
4             $106836.00

In [14]:
df["Department"] = df["Department"].str.title()
df.head()

Name            Position Title        Department  \
0      Aaron,  Elvia J          Water Rate Taker       Water Mgmnt   
1    Aaron,  Jeffery M            Police Officer            Police   
2       Aaron,  Karina            Police Officer            Police   
3  Aaron,  Kimberlei R  Chief Contract Expediter  General Services   
4  Abad Jr,  Vicente M         Civil Engineer Iv       Water Mgmnt   

  Employee Annual Salary  
0              $90744.00  
1              $84450.00  
2              $84450.00  
3              $89880.00  
4             $106836.00

### replace() method

The function is used to replace a string, regex, list, dictionary, series, or number.

In the Department column, "Management“ is shortened to "Mgmnt“ while in the Employee Annual Salary column, there is dollar sign ( $ ) in each values. These data can be cleaned using replace() method.

Firstly, we have to understand how the method works. The first argument is the pattern that we want to replace. The second argument is what we want it to be replaced with.


In [15]:
"Hello World".replace("l", "!")

'He!!o Wor!d'

In [16]:
df["Department"] = df["Department"].str.replace("Mgmnt", "Management")
df.head(3)

Name    Position Title        Department  \
0    Aaron,  Elvia J  Water Rate Taker  Water Management   
1  Aaron,  Jeffery M    Police Officer            Police   
2     Aaron,  Karina    Police Officer            Police   

  Employee Annual Salary  
0              $90744.00  
1              $84450.00  
2              $84450.00

Let's remove the dollar sign in Employee Annual Salary column as well and change the data type to float.

In [17]:
df["Employee Annual Salary"] = df["Employee Annual Salary"].str.replace("$", "").astype(float)
df.head(3)

Name    Position Title        Department  \
0    Aaron,  Elvia J  Water Rate Taker  Water Management   
1  Aaron,  Jeffery M    Police Officer            Police   
2     Aaron,  Karina    Police Officer            Police   

   Employee Annual Salary  
0                 90744.0  
1                 84450.0  
2                 84450.0

Now, we can do mathematical operations on that column. We can get the average annual salary which is 80,204.

In [18]:
df["Employee Annual Salary"].mean()

80204.178633899

### .contains() , endswith() startswith() methods

These methods return Boolean values, either True or False.

Let see how many jobs that are related to water. Out of 32,061, only 111 jobs are related to water.


In [19]:
mask = df["Position Title"].str.contains("Water")
df[mask]

Name                                 Position Title  \
0              Aaron,  Elvia J                               Water Rate Taker   
554         Aluise,  Vincent G             Foreman Of Water Pipe Construction   
671            Ander,  Perry A                               Water Chemist Ii   
685        Anderson,  Andrew J  District Superintendent Of Water Distribution   
702          Anderson,  Donald             Foreman Of Water Pipe Construction   
1054          Ashley,  Karma T                               Water Chemist Ii   
1079         Atkins,  Joanna M                               Water Chemist Ii   
1181        Azeem,  Mohammed A                               Water Chemist Ii   
1285            Bajic,  John A                          Water Meter Machinist   
2400          Bolton,  Brian E                               Water Rate Taker   
2586           Boyce,  Adner L                               Water Chemist Ii   
2745          Brandys,  Daniel                               Water Chemist Ii   
3143          Brown,  Sharon L                               Water Rate Taker   
3246      Buchanan,  Anthony L                               Water Chemist Ii   
3456             Burt,  Carl S                               Water Rate Taker   
3626            Cage,  Ophelia                               Water Rate Taker   
3663     Calderone,  Michael P             Foreman Of Water Pipe Construction   
4889            Clark,  Mark A                          Water Meter Machinist   
5506         Cooper Jr,  Eddie                               Water Chemist Ii   
6337      Davenport,  Clarence                          Water Meter Machinist   
6421       Davis Iii,  Wallace                   Gen Supt Of Water Management   
6564      Decaluwe,  Stanley J             Foreman Of Water Pipe Construction   
6954            Diaz,  Oscar A                               Water Rate Taker   
7356         Dowdy,  Timothy J             Foreman Of Water Pipe Construction   
7463         Driver,  Claude A             Foreman Of Water Pipe Construction   
7641       Durant,  Patricia B                               Water Rate Taker   
7650        Durham,  Patrick L                          Water Meter Machinist   
8092        Espinosa,  Rodolfo                               Water Rate Taker   
8660   Fitzpatrick,  Charles W             Foreman Of Water Pipe Construction   
9325       Gaither,  Rasheda K                            Chief Water Chemist   
...                        ...                                            ...   
23187      Quintanilla,  Edwin                               Water Chemist Ii   
24003         Rios,  Francisco                               Water Rate Taker   
24178         Roberts,  Akilah                               Water Chemist Ii   
24266         Robinson,  Jerry                           Water Meter Assessor   
24422     Rodriguez,  Edward L                           Water Meter Assessor   
24501      Rodriguez,  Marco A                               Water Rate Taker   
24945       Rudmin,  Kenneth D                    Water Quality Inspector I/C   
25055       Russnak,  Thomas W                           Water Meter Assessor   
25233           Saleh,  Hassen         Dir Of Water Purification Laboratories   
25260      Salinas,  Eduardo S                 Engineer Of Water Purification   
26372   Siddiqui,  Abdurrazzaq                              Water Chemist Iii   
26519         Sims,  Demetrius                               Water Rate Taker   
26625           Skowron,  John                               Water Rate Taker   
26871          Smith,  Nancy A                               Water Rate Taker   
27015        Sojka,  Jeffrey A                           Water Meter Assessor   
27368      Stang,  Christopher             Foreman Of Water Pipe Construction   
27537    Stevenson Jr,  Walter             Foreman Of Water Pipe Construction   
27635       Stokes,  Carolyn M                           

We can identify specialists by using **endswith()** method. Here, we have 100 specialists from different departments


In [20]:
mask = df["Position Title"].str.endswith("Specialist")
df[mask]

Name  \
308           Alarcon,  Luis J   
422           Allain,  Carolyn   
705        Anderson,  Edward M   
1163           Ayala Jr,  Juan   
1558       Barrett,  Barbara J   
1869        Beltran,  Mauricio   
2095        Biggane,  Thomas M   
2319            Bobba,  Mahita   
2519      Boston,  Nathaniel K   
2678           Bradley,  Jeena   
4109            Carter,  Alice   
4479             Chan,  Joseph   
5043         Coco,  Michelle T   
5307          Condon,  Kristin   
5633        Cortes,  Jezieel T   
5921         Crump,  Carolyn R   
6136        Czahor,  Patrick J   
6571          Decker,  Grace A   
6599    Degard,  Christopher L   
7317        Dotson,  William L   
7387          Doyle,  George D   
7423         Drain,  Josalyn T   
8753       Flores,  Archibaldo   
8760        Flores,  Eduardo S   
9122          Freelon,  Lisa L   
9366        Gallagher,  John E   
9484        Gapinski,  Frank J   
9562         Garcia,  Joseph E   
10374      Gonzalez,  Lylianis   
10581        Graham,  Kendra D   
...                        ...   
23533     Reckinger,  Jeffry D   
23952          Riley,  Altha S   
24156           Roa,  Edward J   
24525        Rodriguez,  Ralph   
24645          Rollins,  Delia   
25054         Russ,  Johanna M   
25445       Sanders,  Reginald   
25720         Schauf,  Ralph G   
26863        Smith,  Michael L   
26982         Snow,  Michael J   
27306         Sriver,  Nikki K   
27576        Stewart Jr,  John   
27593        Stewart,  Stacy P   
28077        Szymusiak,  Diane   
28353       Terrell,  Lynnette   
28477       Thomas,  Michael E   
28578       Thrower,  Derwin K   
29502     Vaughn Jr,  Joseph T   
30060            Walsh,  Morag   
30156           Ward,  Sonji S   
30287         Watkins,  Gary G   
30914        Williams,  Jana Y   
31118          Willis,  Kirk J   
31296     Wisniewski,  Jerry V   
31320    Wittman,  Elisabeth C   
31653       Yeksigian,  John P   
31667         Yoder,  Teresa G   
31688  Youngbloom,  Laurence G   
31717       Young,  Kimberly M   
31837            Zapata,  Hugo   

                                          Position Title  \
308                           Loan Processing Specialist   
422                 Senior Telecommunications Specialist   
705                            Sr Procurement Specialist   
1163                         Field Sanitation Specialist   
1558                       Technical Training Specialist   
1869                              Procurement Specialist   
2095                                 Security Specialist   
2319                          Grants Research Specialist   
2519                         Field Sanitation Specialist   
2678                          Grants Research Specialist   
4109                                   Safety Specialist   
4479                           Sr Procurement Specialist   
5043                             Investigator Specialist   
5307                               Hr Records Specialist   
5633                           Sr Procurement Specialist   
5921                         Field Sanitation Specialist   
6136                         Field Sanitation Specialist   
6571                               Hr Records Specialist   
6599                           Sr Procurement Specialist   
7317                              Procurement Specialist   
7387                                   Safety Specialist   
7423                         Field Sanitation Specialist   
8753                                   Safety Specialist   
8760                                   Safety Specialist   
9122                           Sr Procurement Specialist   
9366                                   Safety Specialist   
9484                Senior Telecommunications Specialist   
9562              Rehabilitation Construction Specialist   
10374                             Procurement Specialist   
10581                                  Safety Specialist   
...                                                  ... 

We can also do identification by using **startswith()** method. 

In [21]:
mask = df["Position Title"].str.startswith("Water")
df[mask]

Name               Position Title        Department  \
0              Aaron,  Elvia J             Water Rate Taker  Water Management   
671            Ander,  Perry A             Water Chemist Ii  Water Management   
1054          Ashley,  Karma T             Water Chemist Ii  Water Management   
1079         Atkins,  Joanna M             Water Chemist Ii  Water Management   
1181        Azeem,  Mohammed A             Water Chemist Ii  Water Management   
1285            Bajic,  John A        Water Meter Machinist  Water Management   
2400          Bolton,  Brian E             Water Rate Taker  Water Management   
2586           Boyce,  Adner L             Water Chemist Ii  Water Management   
2745          Brandys,  Daniel             Water Chemist Ii  Water Management   
3143          Brown,  Sharon L             Water Rate Taker  Water Management   
3246      Buchanan,  Anthony L             Water Chemist Ii  Water Management   
3456             Burt,  Carl S             Water Rate Taker  Water Management   
3626            Cage,  Ophelia             Water Rate Taker  Water Management   
4889            Clark,  Mark A        Water Meter Machinist  Water Management   
5506         Cooper Jr,  Eddie             Water Chemist Ii  Water Management   
6337      Davenport,  Clarence        Water Meter Machinist  Water Management   
6954            Diaz,  Oscar A             Water Rate Taker  Water Management   
7641       Durant,  Patricia B             Water Rate Taker  Water Management   
7650        Durham,  Patrick L        Water Meter Machinist  Water Management   
8092        Espinosa,  Rodolfo             Water Rate Taker  Water Management   
10170        Godinez,  Julio A        Water Meter Machinist  Water Management   
10457        Gordon,  Donald R        Water Meter Machinist  Water Management   
10707           Green,  John F        Water Meter Machinist  Water Management   
11857          Hedrick,  Lee A             Water Rate Taker  Water Management   
12408      Hjelmgren,  James A        Water Meter Machinist  Water Management   
12487              Ho,  Kwok L  Water Conservation Engineer  Water Management   
12594               Ho,  Nam C             Water Chemist Ii  Water Management   
12781     Howell Jr,  George B             Water Chemist Ii  Water Management   
13361           Jacob,  Lovely            Water Chemist Iii  Water Management   
13476        Jansky,  Thomas A        Water Meter Machinist  Water Management   
...                        ...                          ...               ...   
20936  O Connor Jr,  Michael J        Water Meter Machinist  Water Management   
20962          Odibo,  Grace E             Water Chemist Ii  Water Management   
21078     Olbera Sr,  Martin R        Water Meter Machinist  Water Management   
21212         O Neal,  Rozella         Water Meter Assessor  Water Management   
21788           Parker,  Terry             Water Chemist Ii  Water Management   
22384    Pfannkuche,  Joseph W        Water Meter Machinist  Water Management   
22916         Price,  Johnny L        Water Meter Machinist  Water Management   
23102          Putz,  Andrea R        Water Quality Manager  Water Management   
23187      Quintanilla,  Edwin             Water Chemist Ii  Water Management   
24003         Rios,  Francisco             Water Rate Taker  Water Management   
24178         Roberts,  Akilah             Water Chemist Ii  Water Management   
24266         Robinson,  Jerry         Water Meter Assessor  Water Management   
24422     Rodriguez,  Edward L         Water Meter Assessor  Water Management   
24501      Rodriguez,  Marco A             Water Rate Taker  Water Management   
24945       Rudmin,  Kenneth D  Water Quality Inspector I/C  Water Management   
25055       Russnak,  Thomas W         Water Meter Assessor  Water Management   
26372   Siddiqui,  Abdurrazzaq            Water Chemist Iii  Water Management   
26519         Sims,  Demetrius             Water Rate Tak

### strip() , lstrip() , rstrip()

Strip means removing white space at the start or end of a string.

**lstrip()** stands for left strip. It will only remove white space on the left. On the other hand, **rstrip()** will only remove white space on the right.

These methods are very handy to clean unneeded spaces in the data. Sometimes users tend to make mistakes on input.



In [22]:
"         Hello Word   ".strip()

'Hello Word'

In [23]:
"         Hello Word   ".lstrip()

'Hello Word   '

In [24]:
"         Hello Word   ".rstrip()

'         Hello Word'

In [25]:
df["Name"] = df["Name"].str.strip()
df["Name"].head()

0        Aaron,  Elvia J
1      Aaron,  Jeffery M
2         Aaron,  Karina
3    Aaron,  Kimberlei R
4    Abad Jr,  Vicente M
Name: Name, dtype: object

### Spilit String by Character using .str.split() method

Split string by a character using **.str.split()** method. It will split strings around given separator/delimiter.

The split() method will divide the string based on the separator that we specify. In the example, the separator is a space ( " " )


In [26]:
"My name is Ali bin Abu".split(" ")

['My', 'name', 'is', 'Ali', 'bin', 'Abu']

We can separate the Last name and first name in Name column.

In [27]:
lastName =  df["Name"].str.split(",").str.get(0).str.strip()
lastName

0                 Aaron
1                 Aaron
2                 Aaron
3                 Aaron
4               Abad Jr
5                Abarca
6                Abarca
7               Abascal
8                Abbasi
9            Abbatacola
10          Abbatemarco
11               Abbate
12               Abbott
13               Abbott
14           Abbruzzese
15             Abdallah
16            Abdelhadi
17           Abdellatif
18          Abdelmajeid
19        Abdollahzadeh
20          Abdul-Karim
21             Abdullah
22             Abdullah
23             Abdullah
24             Abdullah
25          Abdulsattar
26         Abdul-Shakur
27           Abdulwahab
28              Abejero
29       Abercrombie Iv
              ...      
32032         Zuniga Jr
32033            Zuniga
32034            Zuniga
32035            Zuniga
32036            Zuniga
32037            Zuniga
32038              Zuno
32039             Zupan
32040          Zupancic
32041          Zurawski
32042          Z

In [28]:
firstName =  df["Name"].str.split(",").str.get(1).str.strip().str.split(" ").str.get(0)
firstName.head(20)

0           Elvia
1         Jeffery
2          Karina
3       Kimberlei
4         Vicente
5          Anabel
6        Emmanuel
7           Reece
8     Christopher
9          Robert
10          James
11          Terry
12          Betty
13         Lynise
14        William
15           Zaid
16      Abdalmahd
17           Aref
18           Aziz
19            Ali
Name: Name, dtype: object

After we have extracted the First Name and Last Name, we can add a new column into the DataFrame using the insert() method.

In [29]:
df.insert(0,column = "First Name" , value = firstName)
df.insert(1,column = "Last Name" ,value = lastName)
df.head()

First Name Last Name                 Name            Position Title  \
0      Elvia     Aaron      Aaron,  Elvia J          Water Rate Taker   
1    Jeffery     Aaron    Aaron,  Jeffery M            Police Officer   
2     Karina     Aaron       Aaron,  Karina            Police Officer   
3  Kimberlei     Aaron  Aaron,  Kimberlei R  Chief Contract Expediter   
4    Vicente   Abad Jr  Abad Jr,  Vicente M         Civil Engineer Iv   

         Department  Employee Annual Salary  
0  Water Management                 90744.0  
1            Police                 84450.0  
2            Police                 84450.0  
3  General Services                 89880.0  
4  Water Management                106836.0

Once the new columns are added, we may drop the Name column to avoid redundancy.

In [30]:
# df.drop(labels= "Name" , axis = 1, inplace=True)
# df.head()

In [31]:
df["First Name"].value_counts().head()

Michael    1153
John        899
James       676
Robert      622
Joseph      537
Name: First Name, dtype: int64

In **split()**, the **expand parameter** will expand the split strings into separate columns.

In [32]:
df["Name"].str.split(",", expand=True).head()

0              1
0    Aaron        Elvia J
1    Aaron      Jeffery M
2    Aaron         Karina
3    Aaron    Kimberlei R
4  Abad Jr      Vicente M

**split()** has **n parameter** which we can use to decide how many occurrences to split. The example shows that we are splitting strings around a space. Since, we give argument of 1 to the n parameter, it will only split the strings around the first space it finds.

In [33]:
df["Position Title"].str.split(" ", expand=True, n = 1).head()

0                   1
0   Water          Rate Taker
1  Police             Officer
2  Police             Officer
3   Chief  Contract Expediter
4   Civil         Engineer Iv

## Regular Expression (ReGex) with Pandas

Regex is a powerful library to search for word(s) in the string. It can also find sequence of characters/words.

Example: Validating a password format. When we want to create a new password, there are some requirements to be fulfilled like it should include uppercase, lowercase, numbers, and unique character.

Python has a library specifically for regex.

Python ReGex Cheat Sheet : https://www.cheatography.com/mutanclan/cheat-sheets/python-regular-expression-regex/

In [34]:
import re

In **re module**, there is a few common libraries used.
 
**match()** function returns a matched object if the text matches the pattern <br>
**search()** function checks for a match anywhere in the string. <br>
**compile()** Compiles a regular expression pattern into a regular expression object. <br>
**findall()** Return all non-overlapping matches of pattern in string, as a list of strings

Tips: When you need to use an expression several times in a single program, using the compile() function to save the resulting regular expression object for reuse is more efficient.

**dot (.) : Match any character except newline**

In [35]:
text = "Amin Hakim"
pattern = r"Ami."

pattern = re.compile(pattern)

result = pattern.search(text)
print(result.group())

Amin


In [36]:
text2 = "Amir Haiqal"

pattern2 = pattern.search(text2)
print(pattern2.group())

Amir


**bracket ([]) : Match anything in the bracket.**

In [37]:
text3 = "I'm 21 years old"

pattern = re.compile("[1-9]")
result = pattern.search(text3)
print(bool(result))

True


In Pandas, we can use regex to search the pattern that we want in the text. As I mentioned before, we can include regex in **str.contains() function**.

For example, we want to search for Position Title which has a digit in it.

In [38]:
df.loc[df["Position Title"].str.contains("[1-9]")].head(10)

First Name     Last Name                    Name  \
275        Mark         Akana          Akana,  Mark S   
892     Gilbert          Arce          Arce,  Gilbert   
1151     Rhonda         Avina        Avina,  Rhonda S   
1233      Gayla       Baggett       Baggett,  Gayla D   
1934    Timothy         Beran         Beran,  Timothy   
2170     Eunice  Bishop Abren   Bishop Abren,  Eunice   
2751    Michael    Bransfield  Bransfield,  Michael P   
3521     Sheila        Butler         Butler,  Sheila   
3561      James         Byrne         Byrne,  James T   
3764      Brian      Campbell      Campbell,  Brian P   

                             Position Title Department  Employee Annual Salary  
275   Police Officer/Explsv Detect K9 Hndlr     Police                 95178.0  
892   Police Officer/Explsv Detect K9 Hndlr     Police                 98016.0  
1151        Communications Operator I - 311       Oemc                 59184.0  
1233        Communications Operator I - 311       Oemc                 62004.0  
1934  Police Officer/Explsv Detect K9 Hndlr     Police                 91752.0  
2170        Communications Operator I - 311       Oemc                 68028.0  
2751  Police Officer/Explsv Detect K9 Hndlr     Police                 98016.0  
3521  Police Officer/Explsv Detect K9 Hndlr     Police                 95178.0  
3561  Police Officer/Explsv Detect K9 Hndlr     Police                 98016.0  
3764  Police Officer/Explsv Detect K9 Hndlr     Police                 95178.0

We can go very specific with additional regex. Say, we want to search for Position Title, which contains the word, "Police" as well as number(s).

- **dot (.)** : Match any character except newline
- **star ( * )** : Zero or more occurrences
- **bracket ([ ])** : Match with anything in the bracket.

In [39]:
df.loc[df["Position Title"].str.contains("Police.*[1-9]")].head(10)

First Name   Last Name                    Name  \
275        Mark       Akana          Akana,  Mark S   
892     Gilbert        Arce          Arce,  Gilbert   
1934    Timothy       Beran         Beran,  Timothy   
2751    Michael  Bransfield  Bransfield,  Michael P   
3521     Sheila      Butler         Butler,  Sheila   
3561      James       Byrne         Byrne,  James T   
3764      Brian    Campbell      Campbell,  Brian P   
5593    Rodolfo    Coronado    Coronado,  Rodolfo E   
6344       Paul   Davenport      Davenport,  Paul B   
6937     Joshua        Diaz           Diaz,  Joshua   

                             Position Title Department  Employee Annual Salary  
275   Police Officer/Explsv Detect K9 Hndlr     Police                 95178.0  
892   Police Officer/Explsv Detect K9 Hndlr     Police                 98016.0  
1934  Police Officer/Explsv Detect K9 Hndlr     Police                 91752.0  
2751  Police Officer/Explsv Detect K9 Hndlr     Police                 98016.0  
3521  Police Officer/Explsv Detect K9 Hndlr     Police                 95178.0  
3561  Police Officer/Explsv Detect K9 Hndlr     Police                 98016.0  
3764  Police Officer/Explsv Detect K9 Hndlr     Police                 95178.0  
5593  Police Officer/Explsv Detect K9 Hndlr     Police                 95178.0  
6344  Police Officer/Explsv Detect K9 Hndlr     Police                 95178.0  
6937  Police Officer/Explsv Detect K9 Hndlr     Police                 88656.0

Let's search for Position Title that starts with **Water** and ends with **Taker**

- **^** : Start of string
- **Bracket []** : Match anything in the bracket. Eg: [Ww] means match any which have w or W
-  **Dot (.)** : Match any single character except the newline character.
- **star ( * )** : Zero or more occurrences.
- **\b** : Word boundary. Spaces between word.


In [40]:
df.loc[df["Position Title"].str.contains("^[wW]ater.*Taker")].head(15)

First Name  Last Name                 Name    Position Title  \
0          Elvia      Aaron      Aaron,  Elvia J  Water Rate Taker   
2400       Brian     Bolton     Bolton,  Brian E  Water Rate Taker   
3143      Sharon      Brown     Brown,  Sharon L  Water Rate Taker   
3456        Carl       Burt        Burt,  Carl S  Water Rate Taker   
3626     Ophelia       Cage       Cage,  Ophelia  Water Rate Taker   
6954       Oscar       Diaz       Diaz,  Oscar A  Water Rate Taker   
7641    Patricia     Durant  Durant,  Patricia B  Water Rate Taker   
8092     Rodolfo   Espinosa   Espinosa,  Rodolfo  Water Rate Taker   
11857        Lee    Hedrick      Hedrick,  Lee A  Water Rate Taker   
13729     Curtis    Johnson   Johnson,  Curtis V  Water Rate Taker   
14771     Inayat       Khan      Khan,  Inayat U  Water Rate Taker   
15215    Anthony  Kordowski  Kordowski,  Anthony  Water Rate Taker   
16189      Byron      Lewis        Lewis,  Byron  Water Rate Taker   
24003  Francisco       Rios     Rios,  Francisco  Water Rate Taker   
24501      Marco  Rodriguez  Rodriguez,  Marco A  Water Rate Taker   

             Department  Employee Annual Salary  
0      Water Management                 90744.0  
2400   Water Management                 78948.0  
3143   Water Management                 82728.0  
3456   Water Management                 90744.0  
3626   Water Management                 90744.0  
6954   Water Management                 90744.0  
7641   Water Management                 90744.0  
8092   Water Management                 78948.0  
11857  Water Management                 78948.0  
13729  Water Management                 75384.0  
14771  Water Management                 90744.0  
15215  Water Management                 86580.0  
16189  Water Management                 90744.0  
24003  Water Management                 78948.0  
24501  Water Management                 90744.0